docs/enhancements/cluster-pool-inventory.md

```diff
@@ -0,0 +1,158 @@
+    clusterDeployments:
+    - name: foo
+    - name: bar
+    - name: baz
+```
+
+(This may seem overengineered, but the idea is to leave room for future expansion.
+For example, perhaps we'll need to specify IP addresses that are tied to each name.
+If we designed the inventory as simply a list of name strings, there would be nowhere to put those IPs.)
+
+When adding a ClusterDeployment, if such an `Inventory` is present, ClusterPool will:
+- Load up the inventory list.
+- Load up the existing pool CDs (it already does this).
+- Subtract the set of names of existing CDs from those in the inventory list.
+- Pick a name from the set that remains and use it for the new CD.
+
+Absent an `Inventory`, ClusterPool will generate a name for new ClusterDeployments as it does today: `{pool name}-{random slug}`
+
+(**Note:** We will not restrict the use of `Inventory` to VSphere.
+I don't know why you would want to use it for other cloud providers, but you can if you like.
+In fact, that's 100% how I plan to test this.)
+
+### How To Use
+For the VSphere case, this allows the administrator to:
+- Preconfigure DNS.
+- Provide the list of (short) hostnames to `ClusterPool.Spec.Inventory`.
+
+For example, if DNS is configured with:
+```
+10.0.0.10  api.foo.example.com
+10.0.0.11  apps.foo.example.com
+10.0.0.12  console.foo.example.com
+
+10.0.1.10  api.bar.example.com
+10.0.1.11  apps.bar.example.com
+10.0.1.12  console.bar.example.com
+```
+the ClusterPool could be configured with:
+```yaml
+...
+spec:
+  baseDomain: example.com
+  inventory:
+    clusterDeployments:
+    - name: foo
+    - name: bar
+  ...
+```
+
+### Validation
+Webhook validation will ensure that, if `Inventory.ClusterDeployments` is specified
+- Each entry contains the `Name` key.
+- The list of `Name` values contains no empties or duplicates.
+
+This validation can be loosened in the future if the need arises without violating the API contract.
+Like maybe for baremetal, we need to supply a preconfigured hardware inventory for each cluster, but it's okay to generate the names.
+
+### `Size` and `MaxSize`
+If `Inventory` is used, `ClusterPool.Spec.Size` and `.MaxSize` are implicitly constrained to the length of the list of nonempty `ClusterDeployments[].Names`.
+Setting either/both to a smaller number still works as you would expect.
+
+### Pool Version
+To make [adding](#adding-an-inventory) and [deleting](#deleting-the-inventory) `Inventory` work sanely, we will adjust the computation of the pool version used for [stale CD detection/replacement](https://issues.redhat.com/browse/HIVE-1058) as follows:
+- When `Inventory` is present, append an arbitrary fixed string before the [final hash operation](https://github.com/openshift/hive/blob/0b9229b91e6f8a3c2bf095efbdf017226e69d026/pkg/controller/clusterpool/clusterpool_controller.go#L479).
+  (We don't want to recalculate the hash any time the inventory changes, as this would treat *all* existing CDs as stale and replace them.)
```

Comment:
here you mean all existing `unclaimed` CDs right? if yes, what is the disadvantage of marking them stale and creating new CDs?